vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd

```diff
@@ -0,0 +1,1007 @@
+encapsulates both a state (the layer's "weights") and a transformation from
+inputs to outputs (a "call", the layer's forward pass).
+
+Here's a densely-connected layer. It has a state: the variables `w` and `b`.
+```{r}
+Linear(keras$layers$Layer) %py_class% {
+  initialize <- function(units = 32, input_dim = 32) {
+    super$initialize()
+    w_init <- tf$random_normal_initializer()
+    self$w <- tf$Variable(
+      initial_value = w_init(
+        shape = shape(input_dim, units),
+        dtype = "float32"
+      ),
+      trainable = TRUE
+    )
+    b_init <- tf$zeros_initializer()
+    self$b <- tf$Variable(
+      initial_value = b_init(shape = shape(units), dtype = "float32"),
+      trainable = TRUE
+    )
+  }
+
+  call <- function(inputs) {
+    tf$matmul(inputs, self$w) + self$b
+  }
+}
+```
+
+You would use a layer by calling it on some tensor input(s), much like a regular
+function.
+
+```{r}
+x <- tf$ones(shape(2, 2))
+linear_layer <- Linear(4, 2)
+y <- linear_layer(x)
+print(y)
+```
+
+`Linear` behaves similarly to a layer present in the python interface to keras
+(e.g., `keras$layers$Dense`).
+
+However, one additional step is needed to make it behave like the builtin layers
+present in the keras R package (e.g., `layer_dense()`).
+
+Keras layers in R are designed to compose nicely with the pipe operator
+(`%>%`), so that the layer instance is conveniently created on demand when an
+existing model or tensor is piped in. In order to make a custom layer similarly
+compose nicely with the pipe, you can call `create_layer_wrapper()` on the layer
+class constructor.
+
+```{r}
+layer_linear <- create_layer_wrapper(Linear)
+```
+
+Now `layer_linear` is a layer constructor that composes nicely with `%>%`, just like the built-in layers:
+
+```{r}
+model <- keras_model_sequential() %>%
+  layer_linear(4, 2)
+
+model(k_ones(c(2, 2)))
+
+model
+```
+
+Because the pattern above is so common, there is a convenience function that
+combines the steps of subclassing `keras$layers$Layer` and calling
+`create_layer_wrapper` on the output: the `Layer` function. The `layer_linear`
+defined below is identical to the `layer_linear` defined above.
+
+```{r}
+layer_linear <- Layer(
+  "Linear",
+  initialize =  function(units = 32, input_dim = 32) {
+    super$initialize()
+    w_init <- tf$random_normal_initializer()
+    self$w <- tf$Variable(initial_value = w_init(shape = shape(input_dim, units),
+                                                 dtype = "float32"),
+                          trainable = TRUE)
+    b_init <- tf$zeros_initializer()
+    self$b <- tf$Variable(initial_value = b_init(shape = shape(units),
+                                                 dtype = "float32"),
+                          trainable = TRUE)
+  },
+
+  call = function(inputs) {
+    tf$matmul(inputs, self$w) + self$b
+  }
+)
+```
+
+For the remainder of this vignette we'll be using the `%py_class%` constructor,
```

Comment:
not sure if these reasons will be convincing to readers ... I guess I'd rather remove that :-)